libs/geometry/src/TangentSpaceMesh.cpp

```diff
@@ -255,16 +353,47 @@ Builder& Builder::algorithm(Algorithm algo) noexcept {
 }
 
 TangentSpaceMesh* Builder::build() {
-    mMesh->mOutput->algorithm = selectAlgorithm(mMesh->mInput);
-    // "Working in progress. Not for use"
-    return NULL;
+    // Work in progress. Not for use.
+    Algorithm algo = selectAlgorithm(mMesh->mInput);
+    MethodPtr method = nullptr;
+    switch (algo) {
+        case Algorithm::FRISVAD:
+            method = [](const TangentSpaceMeshInput* input, TangentSpaceMeshOutput* output) {
+                normalsOnlyMethod(input, output, frisvadKernel);
+            };
+            break;
+        case Algorithm::HUGHES_MOLLER:
+            method = [](const TangentSpaceMeshInput* input, TangentSpaceMeshOutput* output) {
+                normalsOnlyMethod(input, output, hughesMollerKernel);
+            };
+            break;
+        case Algorithm::FLAT_SHADING:
+            method = flatShadingMethod;
+            break;
+        default:
+            break;
+    }
+    assert_invariant(method);
+    method(mMesh->mInput, mMesh->mOutput);
+    return mMesh;
 }
 
 TangentSpaceMesh::TangentSpaceMesh() noexcept
         :mInput(new TangentSpaceMeshInput()), mOutput(new TangentSpaceMeshOutput()) {
 }
 
-TangentSpaceMesh::~TangentSpaceMesh() noexcept {}
+#define CLEAN_HELPER(ptr, inputPtr) \
+   if (ptr && ptr != inputPtr) delete [] ptr;\
+   ptr = nullptr
```

Comment:
I think this should be a function too. Generally we try to avoid macros.

Reply:
done